polaris-service/src/main/java/org/apache/polaris/service/auth/AuthenticatedPolarisPrincipalImpl.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.auth;
+
+import java.util.Set;
+import org.apache.polaris.core.auth.AuthenticatedPolarisPrincipal;
+
+public final class AuthenticatedPolarisPrincipalImpl implements AuthenticatedPolarisPrincipal {
+  private final long id;
+  private final String name;
+  private final Set<String> roles;
+
+  public AuthenticatedPolarisPrincipalImpl(long id, String name, Set<String> roles) {
```

Comment:
I'm actually thinking we should move toward resolving the `PrincipalRole` entities prior to the authorization check. Right now, the Resolver code validates that the `AuthenticatedPrincipal` actually has the grant records to assume the roles specified in the authentication token. That means that we can't decouple the authenticator from the resolver, as someone needs to validate that the roles specified in the token can actually be assumed by the principal. I think that validation should happen at the Authenticator. This allows for third party identity providers to validate the roles in the token without needing grant records, while the default authenticator can continue to rely on the grant records stored by the grant manager.

Reply:
We may have slightly different idea about what the Authenticator should do, I guess :)

From my POV, an Authenticator makes sure the user credentials in the request are valid in the sense that they are not fake or forged (i.e. authentic). I would not necessarily expect the Authenticator to validate that the principal is allowed to assume the roles it wants to assume. I think that is the role of the Authorizer.

In that regard, this PR proposes internal API changes that do decouple resolving principals and and roles against Polaris entities from the Authentication layer.

The main change from my POV is making the entity ID optional in `AuthenticatedPolarisPrincipal`, which is specifically to allow principals that are not mirrored in the Polaris database.

While I think this change is in line with future support for 3rd party identity providers, additional change will be required, I'm sure.

Reply:
I think this is something that was brought up in the Federated Entities doc. Principal Roles are identity roles, not like Catalog Roles, which are groups of privileges. If Principal Roles are a form of identity (e.g., LDAP groups, Okta groups, or IAM group), then membership in those groups is the purview of the Identity Provider, not the Authorizer. I.e., if LDAP says I'm a member of the ENGINEER group, then that claim should be authoritative. The privileges assigned to the ENGINEER group are the purview of the Authorizer (e.g., are there grant records to show the ENGINEER role has access to read a table, is there a policy document that granting the ENGINEER group privileges to create a table in a given namespace...).

If the Authorizer is doubly responsible for validating group membership and for determining privileges, then either the Identity Provider must also be able to store privilege assignments or the identity/group membership must be duplicated in the Authorizer (e.g., an IAM policy must exist that asserts that I'm a member of the ENGINEER group and that the ENGINEER group has the `create_table` privilege).

The main reason I think this strengthens the coupling is that the AuthenticatedPrincipal passes in only the role names without resolving the entities. If we remove the check for the grant records, then a caller could pass in whatever scope they want and be able to assume that role without confirmation they actually have access to that role. 

What I'm hoping to do is change this class to take in the list of validated `PrincipalRole` entities so that the Authenticator itself is responsible for resolving the roles and validating (by whatever means) that the user has access to those roles. At that point, we can change the Resolver and Authorizer to rely on the AuthenticatedPrincipal to directly return the list of `PrincipalRole` entities, rather than just the names, knowing that that list has already been validated by the Authenticator.

Reply:
> If we remove the check for the grant records, then a caller could pass in whatever scope they want and be able to assume that role without confirmation they actually have access to that role.

I do not think so. The caller (Polaris API client) must obtain the credentials first. If the credentials are Polaris-owned, then the authenticator validates that the principal ID is safe to use, but it does not have to load / resolve the roles because it is not know whether the roles are needed for servicing the request. Nonetheless, the authenticity of the caller is established and roles can be validated against Polaris data. If the credentials are 3rd-party, then the IdP must have assigned them (e.g. in JWT claims). Again, the authenticator can validate the claims but does not have to resolve role names to Polaris IDs.

If this we may need another component that deal specifically with mapping external roles / principals to Polaris entities. I would not want to overload the request authentication layer with that duty. This, I think, will complicate integration with external IdPs. For example, authenticating a JWT is in no way connected to Polaris entities. Where entities are synchronously pulled as roles from IdP or via SCIM or some other way, should not affect how credentials are verified to be trusted.

Reply:
> Again, the authenticator can validate the claims but does not have to resolve role names to Polaris IDs.

> We may need another component to deal specifically with mapping external roles / principals to Polaris entities.

This is fine, but right now, that work is being done by the Authenticator + the Resolver and I think that's the wrong separation. The Authenticator today doesn't validate the content of the scopes at all - that's left to the resolver and is only supported by Grant records. This means that it's not enough to add a user to a group in the 3P IdP, but the user must also be added to the group via the Polaris API so that the requisite grant record exists. 

Today, the Polaris `/tokens` endpoint doesn't validate the contents of the `scopes` claim when it generates the token. We can fix this and then the Authenticator doesn't need to do the work of validating the contents of the token. But we'd still need the Resolver to trust the contents of the `AuthenticatedPolarisPrincipal`'s `activePrincipalRoleNames` list and resolve the PrincipalRoles without looking for associated grant records.